Puppet/Daemon.hs

```diff
@@ -99,6 +101,25 @@ initDaemon pref = do
                 templStats
            )
 
+hQueryApis :: Preferences IO -> IO (HieraQueryLayers IO)
+hQueryApis pref = do
+  api0 <- case pref ^. prefHieraPath of
+            Just p  -> startHiera p
+            Nothing -> pure dummyHiera
+  modapis <- getModApis pref
+  pure (api0, modapis)
+
+getModApis :: Preferences IO -> IO (Container (HieraQueryFunc IO))
+getModApis pref = do
+  let ignored_modules = pref^.prefIgnoredmodules
+  dirs <- Directory.listDirectory (pref^.prefPuppetPaths.modulesPath)
+  let
+    modapi :: FoldM IO FilePath (Container (HieraQueryFunc IO))
+    modapi =
+      Foldl.premapM (\m -> (T.pack m, "./modules/" <> m <> "/hiera.yaml"))
+      $ Foldl.prefilterM (\(m,p) -> pure (not (m `elem`ignored_modules)) &&^  Directory.doesFileExist p)
+      $ FoldM (\ s (m,p) -> do h <- startHiera p; pure $ (m,h):s) (pure []) (\l -> pure $ HM.fromList l)
+  Foldl.foldM modapi dirs
```

Comment:
Perhaps the foldl dependency is not that useful? Perhaps something like that could be used (I just wrote this here, it might not even typecheck):

```haskell
getModApis prefs = do
  let ignored_modules = pref ^. prefIgnoredmodules
  dirs <- Directory.listDirectory (pref ^. prefPuppetPaths . modulesPath)
  HM.fromList . catMaybes <$> forM dirs $ \dir -> runMaybeT $ do
    let modulename = T.pack dir
        path = "./modules/" <> m <> "/hiera.yaml"
    guard (modulename `notElem` ignored_modules)
    Directory.doesFileExist path >>= guard
    (modulename,) <$> startHiera path
```